Reference for SAP 4.5 (#498)
mulesoft/docs-connectors: modules/ROOT/pages/sap/sap-reference-45.adoc

| SAP client ID a| String |  Three-digit client number. Leading zeros are preserved if they appear in the number. |  | x
| Destination Name a| String |  Identifier of the destination generated by this connector. If none is set, then a random value is created. |  | 
| Login language a| String |  The language code used by the connector. The language determines the code page used for communicating between the SAP Adapter and the application server. |  `EN` | 
| Extended Properties a| Object | Use to pass in all the database server-specific connection parameters |  | 
| SAP username a| String |  The username the connector uses to log in to SAP |  | x
| SAP password a| String |  The password associated with the login username |  | x
| Application server host a| String |  The host of the SAP application server |  | 
| Message server configuration a| <<MessageServer>> |  The configuration for message server usage |  | 
| Enable RFC tracing on server a| Boolean | Enable or disable RFC trace on the server |  `false` | 
| Enable CPIC tracing on server a| Enumeration, one of:

** INHERIT
** NO_TRACING
** LEVEL_1
** LEVEL_2
** LEVEL_3 |  Enable or disable RFC trace on the server |  | 

> | Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  | 

Reviewer: ```suggestion
| Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, and the test exhausts the associated reconnection strategy without passing, then deployment fails. |  | 
```

Reviewer: I think this preserves the meaning but I'm not sure. What is the thing that's set to true? Reconnection a?

| Connection a| * <<outbound-config_certificate, Certificate>> 
* <<outbound-config_simple-connection-provider, Simple Connection Provider>> 
 | The connection types to provide to this configuration | | x
| Expiration Policy a| <<ExpirationPolicy>> |  Configures the minimum amount of time that a dynamic configuration instance can remain idle before Mule considers it eligible for expiration. This does not mean that the instance expires at the exact moment that it becomes eligible. Mule purges the instances as appropriate. |  | 
| Encoding a| String |  Canonical name of the encoding. See Oracle's supported encodings at https://docs.oracle.com/javase/7/docs/technotes/guides/intl/encoding.doc.html. |  | 
| Default IDoc Version a| String |  Default IDoc version. |  0 | 
| Disable Function Template Cache Flag a| Boolean |  Flag that indicates if the function template cache should be disabled. |  false | 
| Evaluate Response Flag a| Boolean |  Flag that indicates if an error response should throw an exception or let the user handle it as another result. |  false | 
| Log Trace Flag a| Boolean |  Flag that indicates if the trace should be logged on the Mule server. |  false | 

> | Use SAP Format a| Boolean |  This parameter was added to honor the Mule 3 format for returning SAP values. Used in JCoToBusinessObjectParser class when parsing a JCoField.  true - Specifies that values should be returned as strings with the original SAP format. false - Original values are converted to the corresponding Java objects, serialized, and returned.  Default value of PARAM is FALSE. |  false | 

Reviewer: ```suggestion
| Use SAP Format a| Boolean |  This parameter was added to honor the Mule 3 format for returning SAP values. Used in JCoToBusinessObjectParser class when parsing a JCoField.  The value `true` specifies that values are returned as strings with the original SAP format. The value `false` specifies that original values are converted to the corresponding Java objects, serialized, and returned.  Default value of PARAM is FALSE. |  false | 
```

> | Operation Timeout a| Number |  Timeout for operations (in milliseconds; 0 means indefinite waiting) |  0 | 

Reviewer: ```suggestion
| Operation Timeout a| Number |  Timeout for operations in milliseconds. 0 means indefinite waiting. |  0 | 
```

| SAP client ID a| String |  Three-digit client number. Preserve leading zeros if they appear in the number. |  | x
| Destination Name a| String |  Identifier of the destination generated by this connector. If none is set, then a random value is created. |  | 
| Login language a| String |  The language code used by the connector. The language determines the code page used for communicating between the SAP Adapter and the application server. |  `EN` | 
| Extended Properties a| Object |  Use to pass in all the database server-specific connection parameters |  | 
| X.509 Certificate a| String |  Path to the X.509 certificate. This is required when the connector is working as a client. |  | 
| Application server host a| String |  The host of the SAP application server |  | 
| Message server configuration a| <<MessageServer>> |  The configuration for message server usage |  | 
| Enable RFC tracing on server a| Boolean |  Enable or disable RFC trace on the server |  `false` | 
| Enable CPIC tracing on server a| Enumeration, one of:

** INHERIT
** NO_TRACING
** LEVEL_1
** LEVEL_2
** LEVEL_3 |  Enable/disable RFC trace on the server. |  | 

> | Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to true, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  | 

Reviewer: ```suggestion
| Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, and the test exhausts the associated reconnection strategy without passing, then deployment fails. |  | 
```

| SAP client ID a| String |  Three-digit client number. Leading zeros are preserved if they appear in the number |  | x
| Destination Name a| String |  Identifier of the destination generated by this connector. If none is set, then a random value is created. |  | 
| Login language a| String |  The language code used by the connector. The language determines the code page used for communicating between the SAP Adapter and the application server. |  `EN` | 
| Extended Properties a| Object |  Use to pass in all the database server-specific connection parameters |  | 
| SAP username a| String |  The username the connector uses to log in to SAP |  | x
| SAP password a| String |  The password associated with the login username |  | x
| Application server host a| String |  The host of the SAP application server |  | 
| Message server configuration a| <<MessageServer>> |  The configuration for message server usage |  | 
| Enable RFC tracing on server a| Boolean |  Enable or disable RFC trace on the server | `false` | 
| Enable CPIC tracing on server a| Enumeration, one of:

** INHERIT
** NO_TRACING
** LEVEL_1
** LEVEL_2
** LEVEL_3 |  Enable or disable RFC trace on the server |  | 

> | Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  | 

Reviewer: ```suggestion
| Reconnection a| <<Reconnection>> |  When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, and the test exhausts the associated reconnection strategy without passing, then deployment fails. |  | 
```

==== Throws

* SAP:BEGIN_TRANSACTION_ERROR 
* SAP:COMMIT_TRANSACTION_ERROR 
* SAP:CONFIG_ERROR 
* SAP:CONFIRM_TRANSACTION_ERROR 
* SAP:CONNECTION_ERROR 
* SAP:CONNECTIVITY 
* SAP:EXECUTION_ERROR 
* SAP:FIELD_NOT_FOUND_ERROR 
* SAP:INVALID_CREDENTIALS_ERROR 
* SAP:INVALID_HOST_ERROR
* SAP:MISSING_BUSINESS_OBJECT_ERROR 
* SAP:RETRY_EXHAUSTED 
* SAP:ROLLBACK_TRANSACTION_ERROR 
* SAP:TID_CREATION_ERROR 
* SAP:UNKNOWN 
* SAP:XML_PARSING_ERROR  

[[executeSynchronousRemoteFunctionCall]]
=== Execute BAPI function over sRFC
`<sap:execute-synchronous-remote-function-call>`

Executes a BAPI function over a synchronous Remote Function Call (sRFC). A synchronous RFC requires that both the client and server systems are available at the time of communication or data transfer. A synchronous RFC is the most common type and is used when a result is required immediately after the sRFC is executed.  

Use sRFC when communication between source and target systems requires acknowledgments. The resources of the source system wait at the target system and ensure that the message or data is delivered with an acknowledgment (ACKD) and that the data is consistent and reliable for communication.  

> If the target system is not available, the source system resources wait until the target system is available. This may lead to the processes of the source system to go into Sleep/RFC/CPIC Mode at the target systems and block the resources.  

Reviewer: ```suggestion
If the target system is not available, the source system resources wait until the target system is available. This may cause the processes of the source system to go into Sleep/RFC/CPIC Mode at the target systems and block the resources.  
```

sRFC is best used for:  

* Communication between systems 

> * Communication between SAP Web Application Server to SAP GUI 

Reviewer: ```suggestion
* Communication between SAP Web Application Server and SAP GUI 
```

|===

==== Output
[%autowidth.spread]
|===
|Type |Binary
| Attributes Type a| <<SapAttributes>>
|===

=== For Configurations
* <<inbound-config>> 

== Types
[[MessageServer]]
=== Message Server

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Host a| String | The host of the message server |  | x
| System Id a| String | System ID of the SAP system |  | x
| Port a| Number | The port with which the connector logs into the message server |  | 
| Group a| String | Group of SAP application servers |  | 
| Router a| String | SAP router string to use for a system protected by a firewall |  | 
|===

[[Reconnection]]
=== Reconnection

The Reconnection Strategy settings are not enabled. Although disabled, reconnection is automatic when a computer running SAP Connector is disconnected and reconnected. Reconnection occurs based on default conditions established in the SAP JCo libraries. However, if the computer is accessing the SAP instance via VPN, reconnection fails and SAP Connector displays `TIMEOUT` and `CANNOT_SEND_IDOC` errors.

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required

> | Fails Deployment a| Boolean | When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, deployment fails if the test doesn't pass after exhausting the associated reconnection strategy. |  | 

Reviewer: ```suggestion
| Fails Deployment a| Boolean | When the application is deployed, a connectivity test is performed on all connectors. If set to `true`, and the test exhausts the associated reconnection strategy without passing, then deployment fails. |  | 
```

|===

[[reconnect-forever]]
=== Reconnect Forever

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Frequency a| Number | How often in milliseconds to reconnect | |
| blocking |Boolean |If `false`, the reconnection strategy runs in a separate, non-blocking thread | `true` |
|===

[[ExpirationPolicy]]
=== Expiration Policy

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Max Idle Time a| Number | A scalar time value for the maximum amount of time a dynamic configuration instance should be allowed to be idle before it's considered eligible for expiration |  | 
| Time Unit a| Enumeration, one of:

** NANOSECONDS
** MICROSECONDS
** MILLISECONDS
** SECONDS
** MINUTES
** HOURS
** DAYS | A time unit that qualifies the maxIdleTime attribute |  | 
|===

[[SapAttributes]]
=== SAP Attributes

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required
| Transaction Id a| String |  |  | 
|===

[[repeatable-in-memory-stream]]
=== Repeatable In Memory Stream

[%header,cols="20s,25a,30a,15a,10a"]
|===
| Field | Type | Description | Default Value | Required

> | Initial Buffer Size a| Number | The amount of memory that will be allocated to consume the stream and provide random access to it. If the stream contains more data than can be fit into this buffer, then the buffer expands according to the *Buffer size increment* attribute, with an upper limit of *Max in Memory Size*. |  | 

Reviewer: ```suggestion
| Initial Buffer Size a| Number | The amount of memory that will be allocated to consume the stream and provide random access to it. If the data in the stream is larger than this buffer, then the buffer expands according to the *Buffer size increment* attribute, with an upper limit of *Max in Memory Size*. |  | 
```
